File: term_2 (C#, OOP)/homework_4/4.1(ParseTree)/4.1(ParseTree).Tests/ParseTreeTests.cs
namespace ParseTreeNamespace.Tests
{
    using System;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    [TestClass]
    public class ParseTreeTests
    {
        [TestInitialize]
        public void Initialize()
        {
            parseTree = new ParseTree();
        }

        [TestMethod]
        [ExpectedException(typeof(DivideByZeroException))]
        public void EmptуExpressionTest()
        {
            double result = parseTree.Calculate("(/ 1 0)");
        }

        [TestMethod]
        public void UsualExpressionTest()
        {
            double result = parseTree.Calculate("(/ (+ (- 57 67) (/ 36 2)) (/ (* 8 8) (/ (- 80 84) 2)))");
            Assert.AreEqual(result, -0.25);
        }

        private ParseTree parseTree;
    }
}

File: term_4 (F#, FP)/homework_3/homework_3.Tests/ParseTreeTests.fs
namespace homework_3.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open FsUnit.MsTest
open Homework_3.ParseTree

[<TestClass>]
type ParseTreeTests () =

    [<TestMethod>]
    member this.``Eval should calculate parse tree`` () =
        eval <| Multiplication (Addition (Subtraction (Value(8.0), Value(1.0)), Division (Value(3.0), Value(2.0))), Value(2.0))
        |> should equal 17.0
    
    [<TestMethod>]
    member this.``Eval should calculate simple parse tree`` () =
        eval <| Value(2.0)
        |> should equal 2.0
    
    [<TestMethod>]
    member this.``Eval should not crash if divide by zero`` () =
        (fun () -> eval <| Division (Value(3.0), Value(0.0)) |> ignore)
        |> should throw typeof<System.DivideByZeroException>

